Ev/tsh identity

kontsevoy opened the pull request:
## Implementation Notes

First of all, this functionality has already been implemented for `tctl auth sign`. All I had to do was:

* Move it out of `tctl` into `lib/client`
* Hook it up into `client.Login()`
* Make sure `client.Login()` does not "activate" the key (i.e. push it into an active `ssh-agent` or store it into `~/.tsh` if `--out` flag is used).

## Details

Moved the identity generation code code from `tctl` into `lib/client/identity.go`.
Updated teleport client:

`TeleportClient.Login()` used to do too much:

* it retreived the cert and stored it in `client.Key`
* it also generated `ssh.AuthMethod` for that key

Now it simply returns `client.Key` but `client.Key` is now perfectly capable of generating the auth method itself via `ssh.AuthMethod()` method.

When --out is used for `tsh login`, we DO NOT save the keys anywhere but the identity file. This means _nothing_ goes into `~/.tsh`

## Result

This PR closes #1127

kontsevoy committed:
Started working on #1127

Changes:

* Moved identity file code from tctl (because `tctl auth sign` already
  does what we need) into `lib/client/identity.go`
* Updated tctl to use the new stuff.

Next steps:

* Change `tsh` to use that code
* Tests

kontsevoy committed:
Finished implementation of --out flag for tsh login

* Hooked up `MakeIdentityFile()` with `tsh.onLogin()`
* Made sure that `--out` flag prevents from writing into `~/.tsh`

Closes #1127

klizhentas commented on lib/client/api.go:
```diff
@@ -1002,8 +1006,11 @@ func (tc *TeleportClient) Logout() error {
+// If 'activateKey' is true, saves the received session cert into the local
+// keystore (and into the ssh-agent) for future use.
+//
+func (tc *TeleportClient) Login(activateKey bool) (*Key, error) {
```
why would anyone would like not to add the key to the local store? I wonder if we can get rid of this flag and make this always true.

klizhentas commented on lib/client/api.go:
```diff
@@ -968,7 +967,7 @@ func (tc *TeleportClient) ConnectToProxy() (*ProxyClient, error) {
 	// if we get here, it means we failed to authenticate using stored keys
 	// and we need to ask for the login information
-	authMethod, err := tc.Login()
+	key, err := tc.Login(true)
```
not sure why do we return individual key now vs more generic auth methods.

klizhentas commented on lib/client/api.go:
```diff
@@ -1015,7 +1022,7 @@ func (tc *TeleportClient) Login() (*CertAuthMethod, error) {
 	// generate a new keypair. the public key will be signed via proxy if our
 	// password+OTP are legit
-	key, err := tc.MakeKey()
+	key, err := MakeNewKey()
```
Make is redundant, `NewKey` works

kontsevoy replied:
updated.

klizhentas commented on lib/client/api.go:
```diff
@@ -1050,14 +1057,20 @@ func (tc *TeleportClient) Login() (*CertAuthMethod, error) {
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
+	if activateKey {
```
not sure why is this optional

klizhentas commented on lib/client/identity.go:
```diff
@@ -0,0 +1,97 @@
+	const (
+		// the files and the dir will be created with these permissions:
+		fileMode = 0600
+		dirMode  = 0770
```
`why 0770`

kontsevoy replied:
good catch. changed to default `umask` behavior, i.e. 666

klizhentas replied:
tbh I would set modes to 0600 and 0700 because to restrict any external access regardless umask

klizhentas commented on lib/client/identity.go:
```diff
@@ -0,0 +1,97 @@
+		keyPath := username
+
+		// --out flag
+		if fp != "" {
```
`fp` is not very descriptive name, can you make it more descriptive

klizhentas commented on lib/client/identity.go:
```diff
@@ -0,0 +1,97 @@
+package client
```
missing header

kontsevoy committed:
Merge remote-tracking branch 'origin/master' into ev/tsh-identity

kontsevoy committed:
Code review changes

- Added license header to identity.go
- Renamed MakeNewKey() to NewKey()
- Changed the identity file creation to use defaul umask

kontsevoy commented:
retest this please

kontsevoy committed:
Changed identity permissions to be more restrictive

Per @klizhentas suggestion

kontsevoy committed:
Merge remote-tracking branch 'origin/master' into ev/tsh-identity

kontsevoy commented:
@klizhentas changed default mod to `0600` per your suggestion
